expertsystem/state/conservation_rules.py

```diff
@@ -122,129 +120,205 @@ def is_particle_antiparticle_pair(pid1, pid2):
-        pass
+    def get_required_qn_names(self):
+        raise NotImplementedError
 
+    @abstractmethod
     def check_requirements(self, in_edges, out_edges, int_node):
-        for (qn_name_list, condition_functor) in self.get_qn_conditions():
-            # part_props = [x for x in qn_name_list if isinstance(
-            #    x, ParticlePropertyNames)]
-            # if part_props:
-            #    return False
-
-            if not condition_functor.check(
-                qn_name_list, in_edges, out_edges, int_node
+        raise NotImplementedError
+
+
+def rule_conditions(variable_conditions):
+    def decorator(rule_class):
+        all_conditions = []
+        required_qns = []
+        for var_condition in variable_conditions:
+            conditions = None
+            if isinstance(var_condition, tuple):
+                qn_name, conditions = var_condition
+            else:
+                qn_name = var_condition
+            if not (
+                isinstance(
+                    qn_name,
+                    (
+                        StateQuantumNumberNames,
+                        InteractionQuantumNumberNames,
+                        ParticlePropertyNames,
+                        ParticleDecayPropertyNames,
+                    ),
+                )
             ):
-                logging.debug(
-                    "condition %s for quantum numbers %s for rule %s not satisfied",
-                    condition_functor.__class__,
-                    qn_name_list,
-                    self.__class__,
+                raise TypeError(
+                    "qn_name has to be of type "
+                    + "ParticleQuantumNumberNames or "
+                    + "InteractionQuantumNumberNames or "
+                    + "ParticlePropertyNames"
                 )
-                return False
-        return True
+            required_qns.append(qn_name)
+            if conditions:
+                if isinstance(conditions, list):
+                    for condition in conditions:
+                        all_conditions.append(([qn_name], condition))
+                else:
+                    all_conditions.append(([qn_name], conditions))
+
+        def check_requirements(self, in_edges, out_edges, int_node):
+            for (qn_name_list, condition_functor) in all_conditions:
+                if not condition_functor.check(
+                    qn_name_list, in_edges, out_edges, int_node
+                ):
+                    logging.debug(
+                        "condition %s for quantum numbers %s for rule %s not satisfied",
+                        condition_functor.__class__,
+                        qn_name_list,
+                        str(self),
+                    )
+                    return False
+            return True
+
+        def get_required_qn_names(self):  # pylint: disable=unused-argument
+            return required_qns
```

Comment:
Interesting, I had to think about this one. So the `required_qns` is defined and set within the scope of the `decorator`, but then returned in a local function that is then made a method of the decorated class. I'm surprised this works. But I also wonder whether this `get_required_qn_names` can be made into a `property` or `attribute` (called `required_quantum_numbers` or so).

Reply:
That should be possible. I left the interface the same as before as that can be completely refactored in #132. The condition functionality should be improved...